db/similarity.py

```diff
@@ -0,0 +1,193 @@
+        metrics = similarity.utils.init_metrics(force=force)
+        offset = count_similarity()
+        print("Processed {} / {} ({:.3f}%)".format(offset,
+                                                   lowlevel_count,
+                                                   float(offset) / lowlevel_count * 100))
+
+        while count_similarity() < lowlevel_count:
+            batch_query = text("""
+                SELECT id
+                  FROM lowlevel
+              ORDER BY id
+                 LIMIT :batch_size
+                OFFSET :offset
+            """)
+            result = connection.execute(batch_query, {"batch_size": batch_size, "offset": offset})
+            if not result.rowcount:
+                print("Metric {} added for all recordings".format(metric.name))
+                break
+
+            for row in result.fetchall():
+                submit_similarity_by_id(row["id"], metrics=metrics)
+
+            offset = count_similarity()
+            print("Processed {} / {} ({:.3f}%)".format(offset,
+                                                       lowlevel_count,
+                                                       float(offset) / lowlevel_count * 100))
+
+
+def insert_similarity(id, vectors_info):
+    """Inserts a row of similarity vectors for a given lowlevel.id into
+    the similarity table.
+
+        Args: lowlevel.id to be submitted
+              vectors_info, list of tuples of the form:
+              (metric_name, vector, isnan)
+    """
+    with db.engine.connect() as connection:
+        values = []
+        for metric, vector, isnan in vectors_info:
+            value = ('ARRAY' + ('[' + ', '.join(["'NaN'::double precision"] *
+                     len(vector)) + ']' if isnan else str(list(vector))))
+            values.append(value)
+
+        values_string = ', '.join(values)
+        query = text("""
+            INSERT INTO similarity (
+                        id,
+                        mfccs,
+                        mfccsw,
+                        gfccs,
+                        gfccsw,
+                        key,
+                        bpm,
+                        onsetrate,
+                        moods,
+                        instruments,
+                        dortmund,
+                        rosamerica,
+                        tzanetakis)
+                 VALUES (
+                        :id,
+                        %(values)s)
+            ON CONFLICT (id)
+             DO NOTHING
+        """ % {"values": values_string})
+        connection.execute(query, {'id': id})
+
+
+def count_similarity():
+    # Get total number of submissions in similarity table
+    with db.engine.connect() as connection:
+        query = text("""
+            SELECT COUNT(*)
+              FROM similarity
+        """)
+        result = connection.execute(query)
+        return result.fetchone()[0]
+
+
+def submit_similarity_by_id(id, metrics=None):
```

Comment:
@alastair I've refactored and simplified this method, and written some tests for it using mocked classes - I felt like this was a good approach to the tests, but can you confirm? Additionally, this function no longer queries the database itself. Should I move it somewhere like similarity/utils?